extra/crc32_armv8_neon/crc32_armv8.c

```diff
@@ -299,3 +307,35 @@ uint32_t crc32c_aarch64(uint32_t crc, const unsigned char *buffer, uint64_t len)
 
 	return (~crc);
 }
+
+/* There are multiple approaches to calculate crc.
+Approach-1: Process 8 bytes then 4 bytes then 2 bytes and then 1 bytes
+Approach-2: Process 8 bytes and remaining workload using 1 bytes
+Apporach-3: Process 64 bytes at once by issuing 8 crc call and remaining
+            using 8/1 combination.
+
+Based on micro-benchmark testing we found that Approach-2 works best especially
+given small chunk of variable data. */
+unsigned long crc32_aarch64(unsigned long crc32, const unsigned char *buf,
+                             unsigned int len)
+{
+  uint32_t crc= (uint32_t) crc32;
+  const uint64_t *buf8= (const uint64_t *) buf;
+  const uint8_t *buf1= 0;
+
+  crc= ~crc;
+  while (len >= sizeof(uint64_t))
+  {
+    crc= __crc32d(crc, *buf8++);
+    len-= sizeof(uint64_t);
+  }
+
+  buf1= (const uint8_t *) buf8;
+  while (len >= sizeof(uint8_t))
+  {
+    crc= __crc32b(crc, *buf1++);
+    len-= sizeof(uint8_t);
+  }
+
+  return (~crc);
```

Comment:
Please omit the parentheses after `return` and correct the indentation of the parameter list.

Do we really want `unsigned long` for the CRC-32? It is guaranteed to be 32 bits, hence `uint32_t` or `unsigned` or `uint32` would be better. This should of course be fixed accordingly in all implementations.

What if the `buf` is not aligned to 8 bytes? Shouldn’t we first perform a loop of `__crc32b` until it is aligned?

Did you consider unrolling the `__crc32d` loop, like it is done in `ut_crc32_hw()`? Would that improve the performance further?

Reply:
- parentheses omitted.
- unsigned long compatibility came from zlib interface but now that you propose a wrapper around zlib I have corrected it too.
- Also, added additional code to process byte-crc32 before using uin64_t loop.
(I think we are trying to keep asan happy with this but adding additional branches. I presume all new-gen processor can safely handle un-aligned access. bugs if any need to handled on case-by-case).